library/src/main/scala/scala/scalajs/js/Union.scala

```diff
@@ -0,0 +1,65 @@
+/*                     __                                               *\
+**     ________ ___   / /  ___      __ ____  Scala.js API               **
+**    / __/ __// _ | / /  / _ | __ / // __/  (c) 2013-2015, LAMP/EPFL   **
+**  __\ \/ /__/ __ |/ /__/ __ |/_// /_\ \    http://scala-lang.org/     **
+** /____/\___/_/ |_/____/_/ | |__/ /____/                               **
+**                          |/____/                                     **
+\*                                                                      */
+
+
+package scala.scalajs.js
+
+import scala.language.implicitConversions
+
+/** Value of type A or B (union type).
+ *
+ *  Scala does not have union types, but they are important to many
+ *  interoperability scenarios. This type provides a (partial) encoding of
+ *  union types using implicit evidences.
+ */
+@scala.scalajs.js.annotation.RawJSType // Don't do this at home!
+sealed trait |[A, B] // scalastyle:ignore
+
+object | { // scalastyle:ignore
+  /** Evidence that `A <: B`, taking top-level `|`-types into account. */
+  sealed trait Evidence[-A, +B]
+
+  /** A unique (and typically dead-code-eliminated away) instance of
+   *  `Evidence`.
+   */
+  private object ReusableEvidence extends Evidence[scala.Any, scala.Any]
+
+  abstract sealed class EvidenceLowestPrioImplicits {
+    /** If `A <: B2`, then `A <: B1 | B2`. */
+    implicit def right[A, B1, B2](implicit ev: Evidence[A, B2]): Evidence[A, B1 | B2] =
+      ReusableEvidence.asInstanceOf[Evidence[A, B1 | B2]]
+  }
+
+  abstract sealed class EvidenceLowPrioImplicits extends EvidenceLowestPrioImplicits {
+    /** `Int <: Double`, because that's true in Scala.js. */
+    implicit def intDouble: Evidence[Int, Double] =
+      ReusableEvidence.asInstanceOf[Evidence[Int, Double]]
```

Comment:
We could add them, of course, but they're *way* less useful. `intDouble` is important because if your JS library expects a number or a string, it will take `Double | String`. But then if you write `3` it doesn't typecheck ^^ You're unlikely to write `3.toByte` or `3.1f` in that situation, though.

If I want them all I need `Byte -> Short`, `Byte -> Int`, `Byte -> Float`, `Byte -> Double`, `Short -> Int`, `Short -> Float`, `Short -> Double`, `Float -> Double` (because implicits would not magically make the transitive closure in that case). That's *a lot* of implicits to add in the implicit resolution scope (=  longer compile times) for virtually nothing.

Does that sound like a reasonable rationale, or do you want them all?